Fix movies with the same name matching wrong

Razzeee opened the pull request:
@sp1ti

Closes #404

Razzeee commented:
The thing is your problem is collecting right?

We end up with two lists:
a) List of local shows with both versions of blind
b) List of remote/trakt shows only with the newer blind (as it's not added to the collection yet)

We match these. If list b has both blind movies it will match both fine by imdb id. If like in your case it's missing, it's falling back to title search, matches and will never be added as it thinks you already have it.

sp1ti commented:
Yes my issue is with collecting. I followed the matching rules to that point yes but it would still not push one or the other if they have the same title and year despite having a different IMDBid (and thus being different movies). 

A few examples:
"Chaos" (2005) tt0402910 tt0405977
"Brothers" (2015) (on TMDB) tt2835658 tt5803522
"Escape Room" (2017) tt5748392 tt5159414

Razzeee commented:
Yes, same year is still a problem. https://github.com/Razzeee/script.trakt/blob/d9cb536b9d6677e813fbc1482c440409b5faee5e/resources/lib/utilities.py#L104

Not 100% sure what to to about that.

sp1ti commented:
Can the __findInList have filters for "not equal"? If so check that they're not matching the mediaObjectToMatch['ids'].

Razzeee commented:
I've added some more tests and removed the title+year matching. It seems like I inherited it from the old implementation. So getting rid of it is a risk I'm willing to take.

Some testing would be appreciated.

sp1ti commented:
Ok, removing it sort of makes sense I guess 😄  (since people rely on scrappers having those IDs and it is avoids wrong data)). The only issue I see is that trakt from time to time missed some updates and not always has an IMDBid. 
I will test this tomorrow with my data/profile. Thanks 💘

Razzeee commented:
kodi has some big changes with v17 regarding ids. But most scrappers do not support those yet. But whenever that feature lands we can start sending those ids immediately.

sp1ti commented:
Wouldn't this also be missing from the JSON-RPC API? Cannot really see anything else in forms of IDs.

Razzeee commented:
Not sure what your saying.

sp1ti commented:
You are requesting the library data via the JSON-RPC API. But there isn't any field for "IDs" except for `imdbnumber` and `uniqueid`? http://kodi.wiki/view/JSON-RPC_API/v8#VideoLibrary.GetMovies

Razzeee commented:
It's in `uniqueid`. 
See: http://kodi.wiki/view/JSON-RPC_API/v8#Media.UniqueID
And for context https://github.com/xbmc/xbmc/pull/10000

sp1ti commented:
Sorry for the later reply than intended: works well - thank you! It added a bit more than 200 titles. including the example of "Blind" and things like "The Girl on the Train". 
The downside is that now TMDB only movies are added to the list of movies it supposedly adds (went from 153 each cycle to now 724). I'm not representative for the average user but something to keep in mind if this gets implemented.

Also thanks for the uniqueID info; this will hopefully be added to the scrappers around.

Razzeee commented:
Can you explain the TMDB part? 153 to 724 for eg, don't get it right now.

sp1ti commented:
Each sync cycle I have movies shown to me as will be added but never will be.

ex. [Movies Sync] Compared movies, found 158 to add.
[Movies Sync] 158 movie(s) will be added to Trakt.tv collection.

There are a couple of reasons for this: 
- No meta data (neither IMDB or TMDB)
- trakt is missing the movie either because it was not added to TMDB yet OR because the trakt data is outdated and never grabbed an updated IMDB from TMDB
- The movie only has a TMDBid

Because it now won't fall back on year or title matching it presents me with addtional movies it tries to add but cannot actually do for the reasons above

Examples:
https://trakt.tv/movies/death-in-june-behind-the-mask-2005
https://trakt.tv/movies/fuuin-eizou-8-jishou-joou-2012
https://trakt.tv/movies/mushishi-the-next-chapter-drops-of-bells-2015

Razzeee commented:
@rectifyer an the api answer with a reason why something didn't get added?

sp1ti commented:
There is a response that could be logged:
http://docs.trakt.apiary.io/#reference/sync/add-to-collection/add-items-to-collection


In my ideal world the transparency level with the addon would be:
1. A log record for every movie that will be skipped due to missing meta data so that I know which ones need to be added at TMDB (which at the moment will be tried anyways).
2. The current submission log as it is (ideally reinclude the IDs as in the past (I'll try a pull request for this sometime in the future to make it more like the TV log)). This allows me to check what is not picking up from trakt (be it stale IDs or blocked).
3. Logged response body from trakt that is also handled with the guy so that it does not need to lie to me (X added / Y skipped / Z error submitting).

But as I said; I can work around the current limitations as I've been doing for a long time. I'm happy with just having "IDs" respected so everything actually syncs that can and it won't be hooking up the wrong movies together.

rectifyer commented:
@Razzeee It will return in the `not_found` object if it can't match on any of the IDs or title. Check out http://docs.trakt.apiary.io/#reference/sync which explains the order it uses in the **Media objects for syncing** section. Usually it can be fixed on TMDB (and thus next day on Trakt after it syncs), but if needed a Trakt support ticket can be opened for that specific movie and we can look into getting it fixed with the missing IDs.

Razzeee commented:
Sounds like a lot of manual work. :(
Does it also return not found for adault movies that can't be added anyway?

On Mon, 9 Oct 2017, 01:19 Justin, <notifications@github.com> wrote:

> @Razzeee <https://github.com/razzeee> It will return in the not_found
> object if it can't match on any of the IDs or title. Check out
> http://docs.trakt.apiary.io/#reference/sync which explains the order it
> uses in the *Media objects for syncing* section. Usually it can be fixed
> on TMDB (and thus next day on Trakt after it syncs), but if needed a Trakt
> support ticket can be opened for that specific movie and we can look into
> getting it fixed with the missing IDs.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/trakt/script.trakt/pull/406#issuecomment-335046123>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFqyZPXDWRbmeCCRqJxNSVZSxpo4LTnFks5sqVh9gaJpZM4Psvzl>
> .
>

Razzeee commented:
@rectifyer

Razzeee commented on resources/lib/utilities.py:
```diff
@@ -100,15 +100,6 @@ def findMediaObject(mediaObjectToMatch, listToSearch):
     if result is None and 'ids' in mediaObjectToMatch and 'tvdb' in mediaObjectToMatch['ids'] and mediaObjectToMatch['ids']['tvdb']:
         result = __findInList(
             listToSearch, tvdb=mediaObjectToMatch['ids']['tvdb'])
-    # match by title and year it will result in movies with the same title and
```
Better idea, bind this to the setting for name matching fallback?

sp1ti replied:
Forgot about this setting but yes, this would make the most sense!

Razzeee replied:
Same, totally forgot about it. Saw some discussion on another PR yesterday, which got me thinking we might break some usecase when throwing out match by name completely.

Razzeee replied:
Done btw

rectifyer commented:
@Razzeee yes, it would return not found for adult movies as well. Basically anything we can't find a match for in Trakt (badly formed data, invalid id, adult movie, etc).

Razzeee commented:
Feels like I read this different a month ago (or in a hurry...). Thanks for clearing it up.

Razzeee commented:
@sp1ti btw multiple id scrapers have landed, at least for krypton and up. I think only tmdb movies for now, but I'll keep pushing for tvdb.
